Expanding Lr Case plugins

Jt3kt opened the pull request:
Revised, updated, and a little bit expanded.  

New LogRhythm Case functions:
Get-LrCaseEarliestEvidence
Get-LrCasePlaybooks
Get-LrCasePlaybookProcedures
Update-LrCaseEarliestEvidence
Update-LrCaseEarliestEvidenceFromDrilldown
Update-LrCasePlaybookProcedure
Test-LrPlaybookIdFormat
Test-LrUserIdFormat

Jt3kt committed:
Update Update-LrCaseStatus.ps1

Fix typo

Jt3kt committed:
Update CONTRIBUTING.md

Update Slack Channel

Jt3kt committed:
Create Get-LrCaseEarliestEvidence.ps1

Function provides the ability to return date/time of earliest evidence recorded on LogRhythm Case.

Jt3kt committed:
Create Update-LrCaseEarliestEvidenceFromDrilldown.ps1

Function provides the ability to update date/time of earliest evidence recorded on LogRhythm Case based on AIEDrillDown Earliest Evidence.

Jt3kt committed:
Create List-SRFunctions.ps1

Work in progress

Jt3kt committed:
Module planning

Jt3kt committed:
VirusTotal Plugins

Jt3kt committed:
UrlScan Plugin

Jt3kt committed:
To the drawing board with this one.

Jt3kt committed:
Update Preferences to support VirusTotal & UrlScan

Jt3kt committed:
Naming convention fix

LRDeployment -> LrDeployment

Jt3kt committed:
New and updated LR Case functions.

Updated functions associated to CaseEvidence.  Created initial functions for retrieving Case Playbooks and Procedures.

Jt3kt committed:
Create Update-LrCasePlaybookProcedure.ps1

A start, not complete.

Jt3kt committed:
Updated EarliestEvidence functions

Split Update-LrCaseEarliestEvidence from LrCaseEarliestEvidenceFromDrilldown.

Jt3kt committed:
Update Get-LrCaseById.ps1

Fixed typo on 404 error.

Jt3kt committed:
New functions for retrieving playbooks and procedures on cases.

Jt3kt committed:
Remove from LR-Plugins branch.

These will be released as part of their own pull request.

Jt3kt committed:
Consistency changes

Jt3kt committed:
New Case Function: Update-LrCasePlaybookProcedure

Update asignee, status, duedate, notes for a given procedure on a playbook that has been assigned to a case.

Jt3kt committed:
Expanded Procedure parameter options

Now supports procedure by Guid, Name, or Procedure # in the order of procedures.

Jt3kt committed:
Updates to Get-LrList

MaxItemThreshold limit is 1000.  If user doesn't specifiy, it defaults to 1000.  Update to use stored secure credentials.

Jt3kt committed:
Two new functions.  Get-LrLists and Test-LrListType.

Returns a list of all LR Lists.  Can be filtered by name or list type.  Added ListType validator function.

Jt3kt committed:
Updated Get-LrListsGuidByName, now makes use of Get-LrLists

Functions the same but makes use of lower level function, Get-LrLists.

Jt3kt committed:
Create Add-LrListItem.ps1

Work in progress.  Not complete/working at this point.

GeneCupstid commented on src/Public/LogRhythm/Admin/Get-LRListGuidByName.ps1:
```diff
@@ -27,34 +27,23 @@ Function Get-LRListGuidByName {
     [CmdletBinding()]
     Param(
-        [Parameter(Mandatory=$true,Position=0)]
+        [Parameter(Mandatory = $false, Position = 0)]
```
Good catch, the LrList functions were the first LogRhythm cmdlets I added to the module, and I didn't have my style quite nailed down at the time.

GeneCupstid commented on src/Public/LogRhythm/Admin/Get-LRListGuidByName.ps1:
```diff
@@ -27,34 +27,23 @@ Function Get-LRListGuidByName {
+            $Response = Get-LrLists -Name $Name
         }
         catch [System.Net.WebException] {
             $Err = Get-RestErrorMessage $_
```
Because we aren't directly making a Web/Rest request in the function anymore (line 46), we shouldn't use `Get-RestErrorMessage` in the catch block - we should just re-throw the exception from `Get-LrLists`.

Please update the catch block to:

```
$PSCmdlet.ThrowTerminatingError($PSItem)
```

Jt3kt replied:
Yep, I figured that would be the case.  Simple enough change!

GeneCupstid commented on src/Public/LogRhythm/Admin/Get-LrLists.ps1:
```diff
@@ -0,0 +1,132 @@
+        [int] $PageSize,
+
+        [Parameter(Mandatory=$false, Position=4)]
+        [switch] $ValuesOnly
```
**(OPTIONAL)**

There may be value in adding an `Exact` switch to the parameters.  I've done something similar for `Get-LrPlaybooks`.  Because the `Name` parameter is really a regex, it can be a pain to get a specific playbook if other playbooks share a common word.  For example:  "test".  There may be one single list named "test", and a bunch of others including the word "test".  The `Exact` parameter only returns the object that is an exact name match.

```powershell
[Parameter(Mandatory = $false)]
[switch] $Exact
```

Then later...

```powershell
# [Exact] Parameter
# Search "Malware" normally returns both "Malware" and "Malware Options"
# This would only return "Malware"
if ($Exact) {
    $Pattern = "^$Name$"
    $Response | ForEach-Object {
        if(($_.name -match $Pattern) -or ($_.name -eq $Name)) {
            Write-Verbose "[$Me]: Exact playbook name match found."
            $Playbook = $_
            return $Playbook
        }
    }
}
```

GeneCupstid replied:
By the way - It's something I'd like to make uniform for any api calls that use a "name regex".  `Get-LrPlaybooks` was the first time I had thought of it, so there may be some older functions that should use it.  I'll take a look.

Jt3kt replied:
Added this in for Get-LrLists.

GeneCupstid commented on src/Public/LogRhythm/Admin/Get-LrLists.ps1:
```diff
@@ -0,0 +1,132 @@
+        }
+
+        # Process Results
+        if ($ValuesOnly) {
```
`$ValuesOnly` doesn't work because there is no property called `items` in the result.  Probably a cut and paste from `Get-LrList` which returns the list's contents.  Since this function returns multiple lists, `:120-127` should probably be removed?  I don't know if it makes sense to return just the list values from multiple lists at the same time.

Jt3kt replied:
Yea, totally from a cut/paste.  I'll chop it from the function.

GeneCupstid commented on src/Public/LogRhythm/Admin/Get-LrLists.ps1:
```diff
@@ -0,0 +1,132 @@
+        catch [System.Net.WebException] {
+            $Err = Get-RestErrorMessage $_
+            Write-Host "Exception invoking Rest Method: [$($Err.statusCode)]: $($Err.message)" -ForegroundColor Yellow
+            $PSCmdlet.ThrowTerminatingError($PSItem)
```
- Remove `:116`
- Change `:117` to:

```powershell
throw [Exception] "[$Me] [$($Err.statusCode)]: $($Err.message) - $($Err.details) - $($Err.validationErrors)"
```

Related to [Standardize "Catch" blocks for Web Requests #6](https://github.com/SmartResponse-Framework/SmartResponse.Framework/issues/6)

Jt3kt replied:
Done

GeneCupstid commented on src/Public/LogRhythm/Case/Get-LrCaseEarliestEvidence.ps1:
```diff
@@ -0,0 +1,111 @@
+
+        
+        if ($Response -and $Response.earliestEvidence) {
+            if ($Response.earliestEvidence.customDate -ne $null) 
```
`$null` should always be on the left-side when checking equality, according to [PowerShell guidelines](https://github.com/PowerShell/PSScriptAnalyzer/blob/master/RuleDocumentation/PossibleIncorrectComparisonWithNull.md).

Update the equality operations on lines 90 - 99 to the following format:

```powershell
if ($null -ne $Response.earliestEvidence.xxxxx) { ...
```

Jt3kt replied:
Done

GeneCupstid commented on src/Public/LogRhythm/Case/Get-LrCaseEarliestEvidence.ps1:
```diff
@@ -0,0 +1,111 @@
+            if ($Response.earliestEvidence.customDate -ne $null) 
+            {
+                # Custom Date is defined
+                return $Response.earliestEvidence.customDate
```
Let's return these as [datetime] objects.  This will allow for date calculations directly from the output of this cmdlet, which will come in handy in your update functions.

```powershell
if ($null -ne $Response.earliestEvidence.customDate) {
    # Custom Date is defined
    $OutDate = [datetime]::Now
    if ([datetime]::TryParse($Response.earliestEvidence.customDate, [ref]$OutDate)) {
        return $OutDate
    } else {
        return $Response.earliestEvidence.customDate
    }
}
```

GeneCupstid commented on src/Public/LogRhythm/Case/Get-LrCaseEarliestEvidence.ps1:
```diff
@@ -0,0 +1,111 @@
+
+		[Parameter(
+            Mandatory=$true, 
+            Position = 1
```
Add pipeline parameter option before `:44`

```powershell
[Parameter(
    Mandatory=$true, 
    ValueFromPipeline = $true,    # <---- 
    Position = 1
)] 
[object] $Id
```

GeneCupstid replied:
Changed my mind per [Handling Case Id Pararameter #8](https://github.com/SmartResponse-Framework/SmartResponse.Framework/issues/8)

GeneCupstid commented on src/Public/LogRhythm/Case/Get-LrCaseEarliestEvidence.ps1:
```diff
@@ -0,0 +1,111 @@
+        }
+        catch [System.Net.WebException] {
+            $Err = Get-RestErrorMessage $_
+            throw [Exception] "[$Me] [$($Err.statusCode)]: $($Err.message) $($Err.details)`n$($Err.validationErrors)`n"
```
Update `:85` to 

```powershell
throw [Exception] "[$Me] [$($Err.statusCode)]: $($Err.message) - $($Err.details) - $($Err.validationErrors)"
```

GeneCupstid commented on CONTRIBUTING.md:
```diff
@@ -2,7 +2,7 @@
 SmartResponse.Framework has a `master` branch for stable releases and a `develop` branch for daily development. New features and fixes are always submitted to the `develop` branch.
 
-This project follows standard [GitHub flow](https://guides.github.com/introduction/flow/index.html). Please learn and be familiar with how to use Git, how to create a fork of the repository, and how to submit a Pull Request. Contributors are likely willing to help you with using Git if you [ask questions in our slack](https://logrhythmcommunity.slack.com) channel `#smartresponse_framework`.
+This project follows standard [GitHub flow](https://guides.github.com/introduction/flow/index.html). Please learn and be familiar with how to use Git, how to create a fork of the repository, and how to submit a Pull Request. Contributors are likely willing to help you with using Git if you [ask questions in our slack](https://logrhythmcommunity.slack.com) channel `#srf-community`.
```
Good catch!!

GeneCupstid commented on src/Public/LogRhythm/Case/Get-LrCasePlaybooks.ps1:
```diff
@@ -0,0 +1,140 @@
+        $Headers.Add("Authorization", "Bearer $Token")
+        
+        # Retrieve Case Guid
+        $CaseGuid = (Get-LrCaseById -Id $Id).id
```
This is not necessary, as the Rest API will accept either the case id OR the case number.  If the `Id` is invalid, the user will be informed of that by the `catch` below anyway.  If they enter anything other than an `int` or `guid`, `:90` will trap that.

**To simplify, remove `:100` and just use `$Id` instead**

```powershell
# doesn't exist
PS>  Get-LrCaseById -Id 999999
[404]: Case ID 999999 not found, or you do not have permission to view it.

# too big for an int
PS> Get-LrCaseById -Id 9999990023232
Parameter [Id] should be an RFC 4122 formatted string or an integer.
```

Jt3kt replied:
This logic needs to stay, but I've since updated it to be more appropriate.  For this function the CaseGuid needs to be the Case GUID value.  If the Id parameter is passed as an Integer then we need to convert.

`        $IdInfo = Test-LrCaseIdFormat $Id
        if (! $IdInfo.IsValid) {
            throw [ArgumentException] "Parameter [Id] should be an RFC 4122 formatted string or an integer."
        } else {
            # Convert CaseID Into to Guid
            if ($IdInfo.IsGuid -eq $false) {
                # Retrieve Case Guid
                $CaseGuid = (Get-LrCaseById -Id $Id).id
            } else {
                # Id is valid Guid
                $CaseGuid = $Id
            }
        }`

GeneCupstid commented on src/Public/LogRhythm/Case/Get-LrCasePlaybookProcedures.ps1:
```diff
@@ -0,0 +1,155 @@
+        }
+
+        # Get Case Guid
+        $CaseGuid = (Get-LrCaseById -Id $CaseId).id
```
`:101` Getting the case guid should not be necessary, as the Rest API will accept either the case id OR the case number.  If the `Id` is invalid, the user will be informed of that by the `catch` below anyway.  If they enter anything other than an `int` or `guid`, `:96` will trap that.

**To simplify, remove `:100` and just use `$Id` instead**

```powershell
# doesn't exist
PS>  Get-LrCaseById -Id 999999
[404]: Case ID 999999 not found, or you do not have permission to view it.

# too big for an int
PS> Get-LrCaseById -Id 9999990023232
Parameter [Id] should be an RFC 4122 formatted string or an integer.
```

Jt3kt replied:
This logic needs to stay, but I've since updated it to be more appropriate. For this function the CaseGuid needs to be the Case GUID value. If the Id parameter is passed as an Integer then we need to convert.

GeneCupstid commented on src/Public/LogRhythm/Case/Get-LrCasePlaybookProcedures.ps1:
```diff
@@ -0,0 +1,155 @@
+
+        [Parameter(
+            Mandatory = $true,
+            ValueFromPipeline = $true,
```
For now, remove `ValueFromPipeline` on the `$CaseId` parameter.

I think the stronger use case may be to pass the results of `Get-LrCasePlaybooks` to this function.

Jt3kt replied:
Done

GeneCupstid commented on src/Public/LogRhythm/Case/Get-LrCasePlaybookProcedures.ps1:
```diff
@@ -0,0 +1,155 @@
+
+        [Parameter(
+            Mandatory = $false,
+            ValueFromPipeline = $true,
```
Change `ValueFromPipeline` on the `$PlaybookId` parameter to `ValueFromPipelineByPropertyName` and change the name of the parameter to `Id`, so that it matches the output of the Playbook object.

The idea is to allow someone to do the following:

```powershell
PS> Get-LrCasePlaybooks -Id 2068 | Get-LrCasePlaybookProcedures -CaseId 2068
```

Unfortunately the Playbook objects returned by the API for CasePlaybooks does not include the CaseId, so it has to be specified twice.  :(

Jt3kt replied:
Done, for this and for Update-LrCasePlaybookProcedures

GeneCupstid commented on src/Public/LogRhythm/Case/Test-LrPlaybookIdFormat.ps1:
```diff
@@ -0,0 +1,63 @@
+using namespace System.IO
+using namespace System.Collections.Generic
+
+Function Test-LrPlaybookIdFormat {
```
Since PlayBook Id can ONLY be a `guid`, this function can be removed.  

You can use `Test-Guid` anywhere you currently use `Test-LrPlaybookIdFormat`.  The only result of `Test-LrPlaybookIdFormat` would be guid is true or guid is false.  :)

Note: The reason I have `Test-LrCaseIdFormat` is because cases have two kinds of ID - guid and number.  Both are valid ways to look up a unique case, and I wanted to allow the user to specify either with the same parameter.

Jt3kt replied:
Deleted

GeneCupstid commented on src/Public/LogRhythm/Case/Test-LrProcedureIdFormat.ps1:
```diff
@@ -0,0 +1,67 @@
+using namespace System.IO
+using namespace System.Collections.Generic
+
+Function Test-LrProcedureIdFormat {
```
Since Procedure Id can ONLY be a `guid`, this function can be removed.  

You can use `Test-Guid` anywhere you currently use `Test-LrProcedureIdFormat`.  The only result of `Test-LrProcedureIdFormat` would be guid is true or guid is false.  :)

Jt3kt replied:
So this is more of a design Q.  The Test-LrProcedureIdFormat checks if its a Guid, but it also checks if its an integer.  If its a guid, all is good.  If its an integer, then the procedure is being referenced by its position, ie the 4th procedure on playbook X.  Last, if it's not guid or int, then its treated as a string for doing procedure lookup by name match.

Thoughts?

GeneCupstid replied:
Agree with you about this being a design thing.  Let's leave as-is for now.

GeneCupstid commented on src/Public/LogRhythm/Case/Test-LrUserIdFormat.ps1:
```diff
@@ -0,0 +1,62 @@
+using namespace System.IO
+using namespace System.Collections.Generic
+
+Function Test-LrUserIdFormat {
```
User IDs (user number) can only be integers.  It would be fair to allow the user to specify either the name of the user OR the user number.  

Leave this function as-is for now, and I will update `Get-LrUserNumber` to become `Get-LrUser`.  The idea would be that you could pass either an integer or a string to that function, and if it returns a user, you know it's valid.

GeneCupstid commented on src/Public/LogRhythm/Case/Update-LrCaseEarliestEvidence.ps1:
```diff
@@ -0,0 +1,189 @@
+
+        [Parameter(Mandatory = $true, Position = 2)]
+        [ValidateNotNullOrEmpty()]
+        [string] $Timestamp,
```
Change the type of this parameter to `[datetime]`.  That will either accept a real date object, OR turn a string into a [datetime] if it parses correctly.  Then you are guaranteed a valid date and the user gets an error immediately if there's an issue.

Jt3kt replied:
Done

GeneCupstid commented on src/Public/LogRhythm/Case/Update-LrCaseEarliestEvidence.ps1:
```diff
@@ -0,0 +1,189 @@
+        Write-Verbose "[$Me]: Case: $Id EarliestEvidence: $EarliestEvidence"
+
+        # Set Case Creation Date for comparison.  Earliest Evidence !> CaseCreationDate
+        $CaseCreate = (Get-LrCaseById -Id $Id).dateCreated
```
My comment on `Get-LrCaseEarliestEvidence` recommended changing it's output from a `string` to a `datetime`.  This is a great example of why that will be good.   You won't have to convert it, and can immediately do the comparison with the `datetime` provided to this function.

Jt3kt replied:
Updated to return the date/time in the datetime type.

GeneCupstid commented on src/Public/LogRhythm/Case/Update-LrCasePlaybookProcedure.ps1:
```diff
@@ -0,0 +1,357 @@
+                # Validate Playbook Id
+                $PlaybookType = Test-LrPlaybookIdFormat -Id $PlaybookId
+                # Step through array of Playbooks assigned to case looking for match
+                if ($CasePlaybooks -is [array]) {
```
No need to change this, but thought I'd mention it - you don't need to check if `$CasePlaybooks` is an array.  Just go right into a `foreach` loop.  If only one result was returned, the loop will only execute once.  

```powershell
# 2068 only has one playbook assigned
PS> $playbooks = Get-LrCasePlaybooks -Id 2068
PS> foreach($pb in $playbooks2) { $pb.id }
0ECC5442-919C-4D7C-8162-D0BCC689642B
```

GeneCupstid reviewed (changes requested):
Take a look at the comments I've added.  I really appreciate all this work you've done!  It's brought to mind several things I need to take into account as a whole, and I've added several issues to the repo as a result!

Let me know if you have questions or feedback on the comments themselves, and then you can make changes and submit again!

## If you don't have time to make these updates, I don't mind doing them myself - I just want to give you the opportunity first!

GeneCupstid commented:
I didn't actually make any changes, only requested them in the comments - so I'm not sure if you can update this PR with a new set of commits, or if you'll need to submit another PR once  those are complete.

Jt3kt committed:
Function Fixes

Apply round one fixes

Jt3kt committed:
Add exact switch for Get-LrLists function

Jt3kt committed:
Updated Get-LrCaseEarliestEvidence to return datetime format

Jt3kt committed:
Additional fixes and standardization

Updated summary and function info for LrcaseEarliestEvidence.  Removed the BaseUrl and Token from CaseEarliestEvidence, this will be a point to document on issues.  Updated paramater type for DueDate on Update-LrCasePlaybookProcedure to type datetime to fit with standardization.  Found a few $null comparisons that needed to be flipped.

Jt3kt committed:
Update Get-LrCaseEarliestEvidence.ps1

Correct WebException return

Jt3kt committed:
Update Update-LrCasePlaybookProcedure.ps1

A fair amount of fixes and optimizations on this one.

Jt3kt committed:
Update Get-LrCasePlaybookProcedures.ps1

Standardized and optimized.  Worked in recommendations from pull request issue tracker.

Jt3kt committed:
Delete Test-LrPlaybookIdFormat.ps1

No longer needed.

Jt3kt committed:
More standardization

Jt3kt committed:
Updated Add-LrListItem - Partially functional

Jt3kt committed:
Create Test-ValidTCPUDPPort.ps1

Jt3kt committed:
Updated Add-LrListItem

Added support for List:Application Types:Port, PortRange.  Added support for List: IP.   List:IPRange in progress.

Jt3kt committed:
Update Get-LRListGuidByName.ps1

Identify use cases where you would not want to use Exact on LrListGuidByName.  Possibly set the default to true.

Jt3kt committed:
Updated Parameter Identity to Name

Effort to standardize List functions.   Added Exact switch, can likely reduce that.

Jt3kt committed:
Create Get-LrListItems.ps1

Returns powershell object for all list items on a list.

Jt3kt committed:
Create Test-LrListValue.ps1

Identifies if a value is a member of a list.  Provides PSObject with item and list data.

Jt3kt committed:
Update Get-LrCasePlaybooks.ps1

Updated means to retrieve Case Guid.

Jt3kt committed:
Updated Function Synopsis blocks

Jt3kt committed:
Valid release of Add-LrListItem

Appropriately adds item to list.  Handles type inspection, duplicate inspection, and provides PSCustonObject on error.  This enables the passing of multiple values through pipeline without interrupting the processing chain.

Jt3kt committed:
AdminAPI - Get-LrHostDetails, GetLrHostIdentifiers, Get-LrHosts

Three new functions to support retrieving Host information from the LogRhythm Entity structure.

Jt3kt committed:
Create Get-LrIdentities.ps1

Functional cmdlet for returning PSCustomObject of identities.  Does support searching.  Currently does not support the -exact flag.  Synopsis section needs update.

Jt3kt committed:
New function, add file to LR Case

Does not support adding notes in conjunction with file.  Does support uploading more than txt files.  Further testing and documentation required.
